RFCs/0017-TestCase-FullyQualifiedName.md

```diff
@@ -1,49 +1,97 @@
+For example, in the case of source-based test discovery, we need to be able to identify tests from source code. Those same tests also need to be matched to tests discovered by reflection. Without unique and deterministic FQN identifiers, we don't have a reliable way to match corresponding tests. In other cases, such as the hierarchy view or dynamic source location fetching, we need to be able to reliably connect the source representation of a test with the type & method (via reflection) of a test. With this standard, FQN enables us to make those connections. NOTE: Our intent is that the test adapters take responsibility for source-based test discovery. Once this transition is complete, the need to keep the FQNs in sync between discovery methods will be reduced to only making sure the types and methods match up.
 
-## Normalized Form
-When referencing test cases by FullyQualifiedName, the test platform recognizes a normalized form of the FullyQualifiedName. This normalized form is the FullyQualifiedName with segments in parentheses and segments in angle brackets removed. The normalized form will not be unique among TestCases.
+## Requirements of `FullyQualifiedName` and TestCase `ID`
 
-| Fully Qualified Name | Normalized Fully Qualified Name |
-|---|---|
-| `Namespace.Namespace.Class.Method()` | `Namespace.Namespace.Class.Method` |
-| `Namespace.Namespace.Class<Int32>.Method()` | `Namespace.Namespace.Class.Method` |
-| `Namespace.Namespace.Class<Int32>.Method<String>()` | `Namespace.Namespace.Class.Method` |
-| `Namespace.Namespace.Class+InnerClass.Method` | `Namespace.Namespace.Class+InnerClass.Method` |
-| `Namespace.Namespace.Class.Method(x: 1, msg: "string")` | `Namespace.Namespace.Class.Method` |
-| `Namespace.Namespace.Class.Method(1, "string")` | `Namespace.Namespace.Class.Method` |
-| `Namespace.Namespace.Class(init: "Value").Method(param: 23)` | `Namespace.Namespace.Class.Method` |
-| `Namespace.Namespace.Class<String>("Value").Method(23)` | `Namespace.Namespace.Class.Method` |
+* `FullyQualifiedName` should uniquely identify a method that represents test cases (with the optional application of a dataset). There may be duplicate FQNs within a project due to some methods being associated with multiple test cases. Note that it is possible for multiple assemblies, or multiple test executors to use the same FQNs. When comparing FQNs within a Visual Studio session, `Source`, `FQN`, and `ExecutorURI` must be combined to guarantee uniqueness.
+* Each TestCase has an `ID` field that takes the form of a GUID. In the past, this field has been optional. It will now become required. The required `ID` field must be unique for a given test assembly. This `ID` field must also be deterministic, in the sense that a test with the same class, method and method arguments should have the same id. We envision generating this `ID` by hashing the test's class, method, method signature and method arguments into the GUID. 
```

Comment:
The spec should probably clarify whether ID is unique across a given project or whether it can be unique across the solution i.e. whether or not TestCase.Source (and  TestCase.ExecutorUri) are included in the ID generation.

Personally I think having unique IDs across the solution would be great because that means any code that needs to keep track of TestCases can just hold them in a dictionary with ID as key. I just fixed an issue in LUT where the code was incorrectly assuming that it is safe to do this and fails because currently this not safe. It would be great if all code consuming TestCases could just use the ID as key without having to create an intermediate identifier type (comprised of ID, Source and ExecutroUri) for this. This would also make all code that deals with large dictionary of tests (the test store in TW, similar stores in SBD, LUT etc.) more efficient because storage and retrieval of TestCases can be performed by comparing just a single GUID (instead of comparing a GUID and two other strings).

Reply:
I agree this could make the implementation easier in certain ways. I have some concerns about this though.

1. Making the ID globally unique would require that the adapter & framework know all of the data required to make that ID unique including the Source and the ExecutorUri. Given that discovery is abstracted and implemented by the test framework core, does the discovery engine always have access to the source (the assembly file name) and executor? I'm not sure if it would in every case. What about scenarios that are not based on assemblies?
1. In the future, if additional scenarios are added that would allow for a single test case to have multiple "instances" within a single source assembly, it might be very difficult to support if we also require the test frameworks to know the scope of uniqueness of each test case.

It seems more consistent to scope the uniqueness guarantee of test case ids to be within the discovery container. We can always add another deterministic hashing layer at the TestWindow level that further scopes the ids by Source/TFM, and Executor, etc...